modules/systemd-journal/journal-reader.c

```diff
@@ -797,6 +815,8 @@ _free(LogPipe *s)
   JournalReader *self = (JournalReader *) s;
   log_pipe_unref(self->control);
   log_source_free(&self->super.super);
+
+  used_namespaces = g_list_delete_link(used_namespaces, g_list_find(used_namespaces, self->persist_name));
```

Comment:
This breaks the reload mechanism `syslog-ng-ctl reload`.

The general rule of thumb is the following:
- `new()-free()` and `init()-deinit()` should be symmetric in terms of resource allocation/release (not always, but see next point)
- `deinit()` should be somewhat idempotent, because the reload mechanism can do tricky things:
  - `syslog-ng-ctl reload` => `deinit()` old config pipes => `init()` new config pipes
    - `if (new_config_init)` => `free()` old config pipes => reload succeeded
    - `if (!new_config_init)` => `deinit()` half-initialized new config pipes => `init()` old config pipes again => config reverted back to the old one


Reproduction steps:
```
source {
  systemd-journal();
  systemd-journal(namespace("asd"));
};
```

`sbin/syslog-ng-ctl reload` will SIGQUIT:
```
systemd-journal namespace already in use; namespace='asd'
The configuration must not contain more than one systemd-journal() source with the same namespace() option;

Error initializing new configuration, reverting to old config;

systemd-journal namespace already in use; namespace='asd'
The configuration must not contain more than one systemd-journal() source with the same namespace() option;
```

The SIGQUIT happens because the config revert fails, and syslog-ng kills itself in case neither the new nor the old config can be restored.


My suggestion would be moving everything into `init()` and `deinit()`, but making sure that calling `deinit()` twice won't mess up `used_namespaces` or cause a crash.

Reply:
Thank you for the review!

I have moved the code that removes a used namespace from the `used_namespaces` list to the `_deinit()` function. I have tested and the reload mechanism is now working.

There was also an error in the removal code that it searched for `self->persist_name` instead of `self->options->namespace` which I have now fixed.

Reply:
Thanks.